linked_list/linked_list.py

```diff
@@ -9,72 +9,162 @@ def __init__(self, value, next_node = None):
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(1)
+    # Space Complexity: 0(n)
     def add_first(self, value):
-        pass
+        node = Node(value, next_node=self.head)
+        self.head = node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(1)
     def search(self, value):
-        pass
+        current = self.head
+
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        
+        return False 
+
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(1)
     def length(self):
-        pass
+
+        count = 0
+
+        current = self.head
+        
+        while current:
+            
+            current = current.next
+            count += 1
+
+        return count 
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(1)
     def get_at_index(self, index):
-        pass
+
+        if self.head == None:
+            return None
+
+        if index == 0:
+            return self.head.value
+
+        current = self.head
+        counter = 0
+
+        # index - 1 --> we are manipulating the pointer for the previous node
+        while current.next and counter < index - 1:
+            current = current.next
+            counter += 1
+        
+        if current.next:
+            return current.next.value 
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(1)
     def get_last(self):
-        pass
+
+        if self.head == None:
+            return None
+        
+        current = self.head
+        while current.next: 
+            current = current.next
+        
+        return current.value
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(1)
     def add_last(self, value):
-        pass
+
+        if self.head == None:
+            self.head = Node(value, None)
+            return 
+        
+        current = self.head
+
+        while current.next:
+            current.next

```

Comment:
This is why the tests hang

```suggestion
            current = current.next

```